File: bindings/elixir/examples/tool_discovery.exs
#!/usr/bin/env elixir

# Tool Discovery Example for UMICP Elixir SDK

Mix.install([{:jason, "~> 1.4"}])

Code.prepend_path("_build/dev/lib/umicp/ebin")

defmodule ExampleSearchService do
  @moduledoc """
  Example service implementing tool discovery.
  """

  use UMICP.Discovery.DiscoverableService

  alias UMICP.Discovery.{OperationSchema, ServerInfo}

  @impl true
  def list_operations do
    [
      OperationSchema.new(
        name: "search",
        input_schema: %{
          "type" => "object",
          "properties" => %{
            "query" => %{"type" => "string", "description" => "Search query"},
            "max_results" => %{"type" => "integer", "default" => 10}
          },
          "required" => ["query"]
        }
      )
      |> OperationSchema.with_title("Search Operation")
      |> OperationSchema.with_description("Performs a semantic search"),
      OperationSchema.new(
        name: "get_document",
        input_schema: %{
          "type" => "object",
          "properties" => %{
            "document_id" => %{"type" => "string"}
          },
          "required" => ["document_id"]
        }
      )
      |> OperationSchema.with_title("Get Document")
    ]
  end

  @impl true
  def server_info do
    ServerInfo.new(
      server: "example-search-service",
      version: "1.0.0",
      protocol: "umicp"
    )
    |> ServerInfo.with_features(["tool_discovery", "native_types", "mcp_compatible"])
    |> ServerInfo.with_operations_count(2)
    |> ServerInfo.with_mcp_compatible(true)
    |> ServerInfo.with_metadata(%{
      "platform" => "elixir",
      "sdk_version" => UMICP.version()
    })
  end
end

defmodule ToolDiscoveryExample do
  alias UMICP.Discovery.DiscoverableService

  def run do
    IO.puts("\n=== Tool Discovery Example ===\n")

    service = ExampleSearchService

    # List all operations
    IO.puts("Available Operations:")
    operations = service.list_operations()

    for op <- operations do
      IO.puts("  - #{op.name}: #{op.title || "No title"}")

      if op.description do
        IO.puts("    Description: #{op.description}")
      end
    end

    # Get specific schema
    IO.puts("\nDetailed Schema for 'search':")
    schema = service.get_schema("search")
    IO.puts("  Name: #{schema.name}")
    IO.puts("  Title: #{schema.title}")
    IO.puts("  Description: #{schema.description}")

    # Get server info
    IO.puts("\nServer Information:")
    info = service.server_info()
    IO.puts("  Server: #{info.server}")
    IO.puts("  Version: #{info.version}")
    IO.puts("  Protocol: #{info.protocol}")
    IO.puts("  Features: #{Enum.join(info.features, ", ")}")
    IO.puts("  Operations Count: #{info.operations_count}")
    IO.puts("  MCP Compatible: #{info.mcp_compatible}")

    # Generate JSON responses
    IO.puts("\nJSON Responses:")

    {:ok, ops_json} = DiscoverableService.generate_operations_response(service)
    IO.puts("\nList Operations Response:")
    IO.puts(ops_json)

    {:ok, info_json} = DiscoverableService.generate_server_info_response(service)
    IO.puts("\nServer Info Response:")
    IO.puts(info_json)

    IO.puts("\n✅ Tool discovery example completed!")
  end
end

ToolDiscoveryExample.run()

File: bindings/swift/Examples/ToolDiscovery.swift
import Foundation
import UMICP

// Example service implementing tool discovery
class ExampleSearchService: DiscoverableService {
    func listOperations() -> [OperationSchema] {
        return [
            OperationSchema(
                name: "search",
                inputSchema: [
                    "type": AnyCodable("object"),
                    "properties": AnyCodable([
                        "query": ["type": "string", "description": "Search query"],
                        "max_results": ["type": "integer", "default": 10],
                        "include_metadata": ["type": "boolean", "default": false]
                    ]),
                    "required": AnyCodable(["query"])
                ],
                title: "Search Operation",
                description: "Performs a semantic search across documents"
            ),
            OperationSchema(
                name: "get_document",
                inputSchema: [
                    "type": AnyCodable("object"),
                    "properties": AnyCodable([
                        "document_id": ["type": "string", "description": "Document ID"]
                    ]),
                    "required": AnyCodable(["document_id"])
                ],
                title: "Get Document",
                description: "Retrieves a specific document by ID"
            )
        ]
    }

    func serverInfo() -> ServerInfo {
        return ServerInfo(
            server: "example-search-service",
            version: "1.0.0",
            protocol: "umicp",
            features: ["tool_discovery", "native_types", "mcp_compatible"],
            operationsCount: 2,
            mcpCompatible: true
        )
        .withMetadata([
            "platform": AnyCodable("swift"),
            "sdk_version": AnyCodable(UMICP.version)
        ])
    }
}

// Example usage
func toolDiscoveryExample() throws {
    print("=== Tool Discovery Example ===\n")

    let service = ExampleSearchService()

    // List all operations
    print("Available Operations:")
    let operations = service.listOperations()
    for op in operations {
        print("  - \(op.name): \(op.title ?? "No title")")
        if let desc = op.description {
            print("    Description: \(desc)")
        }
    }

    // Get specific schema
    print("\nDetailed Schema for 'search':")
    if let schema = service.getSchema(name: "search") {
        print("  Name: \(schema.name)")
        print("  Title: \(schema.title ?? "none")")
        print("  Description: \(schema.description ?? "none")")
        print("  Input Schema: \(schema.inputSchema)")
    }

    // Get server info
    print("\nServer Information:")
    let info = service.serverInfo()
    print("  Server: \(info.server)")
    print("  Version: \(info.version)")
    print("  Protocol: \(info.protocol)")
    print("  Features: \(info.features?.joined(separator: ", ") ?? "none")")
    print("  Operations Count: \(info.operationsCount ?? 0)")
    print("  MCP Compatible: \(info.mcpCompatible ?? false)")

    // Generate JSON responses
    print("\nJSON Responses:")

    let opsJson = try DiscoveryHelpers.generateOperationsResponse(service: service)
    print("\nList Operations Response:")
    print(opsJson)

    let schemaJson = try DiscoveryHelpers.generateSchemaResponse(
        service: service,
        operationName: "search"
    )
    print("\nGet Schema Response:")
    print(schemaJson)

    let infoJson = try DiscoveryHelpers.generateServerInfoResponse(service: service)
    print("\nServer Info Response:")
    print(infoJson)
}

// Run example
do {
    try toolDiscoveryExample()
    print("\n✅ Tool discovery example completed!")
} catch {
    print("❌ Error: \(error)")
}

